Add experts documentation

ernestum opened the pull request:
## Description

This PR documents how experts can be loaded from disk and from HuggingFace. It also describes how to push your own expert policies to HuggingFace.

Note: this PR branchess off #565 and should be rebased on master after #565 as been merged.

dfilan commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+
+The `huggingface-sb3 package <https://github.com/huggingface/huggingface_sb3>`_ provides
+utilities to push your own models to HuggingFace and load them from there.
+Make sure use the naming scheme helpers
```
Should be "Make sure to use"

ernestum replied:
thanks, fixed

dfilan commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+the policy type.
+By default, the policies are loaded from the HumanCompatibleAI repository, but you can
+override this by setting the `organization` parameter in the `loader_kwargs`.
+When using the python API, you also have to speciyf the environment name as `env_name`.
```
should be specify

ernestum replied:
thanks, fixed

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+
+In the :doc:`../getting-started/first-steps` tutorial, we first trained an expert and
+then imitate it's behavior using :doc:`../algorithms/bc`.
+In practice, you want to load a pre-trained policy for performance reasons.
```
"you want" is a bit strong -- we don't know the user's situation. Could also soften it to "you typicaly want" etc.
```suggestion
In practice, you may want to load a pre-trained policy for performance reasons.
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+then imitate it's behavior using :doc:`../algorithms/bc`.
+In practice, you want to load a pre-trained policy for performance reasons.
+
+The python interface provides a :func:`~imitation.policies.serialize.load_policy`
```
```suggestion
The Python interface provides a :func:`~imitation.policies.serialize.load_policy`
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+In practice, you want to load a pre-trained policy for performance reasons.
+
+The python interface provides a :func:`~imitation.policies.serialize.load_policy`
+function to which you pass a `policy type` and any extra kwargs to pass to the
```
```suggestion
function to which you pass a `policy_type` and any extra kwargs to pass to the
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+----------------------------
+
+To load a policy from disk, use either `ppo` or `sac` as the policy type.
+The path is specified as `path` to the `loader_kwargs` and it should either point
```
```suggestion
The path is specified by `path` in the `loader_kwargs` and it should either point
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,50 @@
+the policy type.
+By default, the policies are loaded from the HumanCompatibleAI repository, but you can
+override this by setting the `organization` parameter in the `loader_kwargs`.
+When using the python API, you also have to specify the environment name as `env_name`.
```
```suggestion
When using the Python API, you also have to specify the environment name as `env_name`.
```

AdamGleave reviewed (commented):
The documentation as written looks good apart from a few minor comments, but it seems to be missing a bit. It only describes how to use it to brain BC/DAgger with experts, but can't we also sample demonstrations from pre-trained experts with GAIL/AIRL (and perhaps density and MCE)? The latter use case seems more common so it'd be great to have documentation for that too, although if you wanted to split that into a separate PR I wouldn't object.

ernestum commented:
We don't have scripts for density and MCE and they need demonstrations (not an expert).
For GAIL/AIRL, if no demonstrations are specified, they are sampled from an expert ad-hoc. I will mention this and add some examples.

ernestum commented:
I added a proper introduction, addressed the case with AIRL/GAIL and added a section on how to register custom experts.
Copilot was surprisingly good at generating usage examples so we have them too.
The new `config_hook` in `expert.py` makes the `loader_kwargs` API somewhat discoverable without digging into the documentation.

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,125 @@
+
+The algorithms in the imitation library are all about learning from some kind of
+expert.
+In many cases it is practical that this expert is a piece of software itself.
```
"practical that" sounds a bit off to my ears, possible rephrasing below, there might be others that better capture your intention.
```suggestion
In many cases it is simplest if this expert is a piece of software itself.
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,125 @@
+learning library.
+
+For example, BC and DAgger can learn from an expert policy and the command line
+interface to AIRL/GAIL allows to specify an expert to sample demonstrations from.
```
```suggestion
interface of AIRL/GAIL allows one to specify an expert to sample demonstrations from.
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,125 @@
+to a zip file containing the policy or a directory containing a `model.zip` file that
+was created by stable-baselines3.
+
+In command line interface the `expert.policy_type` and `expert.loader_kwargs`
```
```suggestion
In the command line interface the `expert.policy_type` and `expert.loader_kwargs`
```

AdamGleave commented on docs/experts/loading-experts.rst:
```diff
@@ -0,0 +1,125 @@
+`rl-baselines3-zoo <https://github.com/DLR-RM/rl-baselines3-zoo/>`_.
+
+
+Custom expert types
```
Nice, hadn't thought to document this but it's an important part of the API :)

AdamGleave commented on src/imitation/scripts/common/expert.py:
```diff
@@ -9,20 +9,37 @@
-def get_expert_policy(venv, policy_type, loader_kwargs, common):
-    if "huggingface" in policy_type:
+@expert_ingredient.config_hook
+def config_hook(config, command_name, logger):
```
Nice, this definitely makes things more self-documenting :)

AdamGleave reviewed (approved):
LGTM, this is much more comprehensive! I've added some minor suggestions re: phrasing. Feel free to merge after addressing those.
